sources/api/netdog/src/networkd/config/network.rs

```diff
@@ -0,0 +1,81 @@
+use networkd_derive::{NetworkDFile, NetworkDSection};
+
+#[derive(Debug, Default, NetworkDFile)]
+pub(crate) struct NetworkConfig {
+    match_section: Option<MatchSection>,
+    link_section: Option<LinkSection>,
+    network_section: Option<NetworkSection>,
+    route_sections: Vec<RouteSection>,
+    dhcp4_section: Option<Dhcp4Section>,
+    dhcp6_section: Option<Dhcp6Section>,
+}
+
+#[derive(Debug, Default, NetworkDSection)]
+#[networkd(section = "Match")]
+struct MatchSection {
+    #[networkd(entry = "Name")]
+    name: Option<String>,
+    #[networkd(entry = "PermanentMACAddress")]
+    permanent_mac_address: Option<String>,
```

Comment:
> We validate input at the time we deserialize network config (net.toml), so at the point we are creating these files we can assume input is valid.

Strong typing at this layer is not about user input validation; rather, it's about making use of the type system so that the compiler has our back when we go to modify this code.

For example, `networkd` has its own rules around fields like `RequiredFamily` that don't map exactly to what's coming from input, so there's value in ensuring that the `required_family` field only gets appropriate values.

The current approach also doesn't seem aligned with how the `wicked` structs work, which do make use of stronger types.

> The macro would probably need to grow to understand annotations for the string representation of each of the enum variants so we would print ipv4, both, any, etc.

Can't the macro just assume that each field implements `Display` or some other trait, and call that impl?

> All that being said, if we intend to publish this macro/structs as a standalone crate somewhere I would absolutely agree about the validation of types.

I don't think it's a single standalone crate; rather, you have:
* macros to simplify serializing Rust structs as systemd unit files
* structs that correspond specifically to the data shapes that networkd expects